[jit] Have ScriptModule inherit from Module

zdevito opened the pull request:
This is accomplished by created replacement _parameters, _buffers,
  and _modules which implement the OrderedDict APIs but which
  actually get/set their members inside script::Module

* Merge TracedModule with ScriptModule
* Move logic of attribute handling into Python bindings rather than
  make script::Module handle it. This was redundant with nn.Module,
  which already handles attribute.

Make TracedModule a subclass of ScriptModule

Move handling of attribute kind logic into bindings.

Port back in TracedModule functionality to ScriptModule

zdevito committed:
* Have ScriptModule inherit from Module
  This is accomplished by created replacement _parameters, _buffers,
  and _modules which implement the OrderedDict APIs but which
  actually get/set their members inside script::Module

* Merge TracedModule with ScriptModule
* Move logic of attribute handling into Python bindings rather than
  make script::Module handle it. This was redundant with nn.Module,
  which already handles attribute.

Make TracedModule a subclass of ScriptModule

Move handling of attribute kind logic into bindings.

Port back in TracedModule functionality to ScriptModule

zdevito commented:
@jamesr66a @apaszke 

This needs a few more tests, I can use some advice on what functionality of nn.Module we should test for.

zdevito committed:
Allow ScriptModule to contain non-script module submodules.

apaszke commented on torch/csrc/jit/script/init.h:
```diff
@@ -6,6 +6,11 @@ namespace torch {
+std::shared_ptr<Graph> createGraphByTracing(
+        py::function func,
+        std::vector<tracer::TraceInput> inputs,
+        size_t num_inputs);
```
Can we move this to `tracer.cpp`? `init.h` shouldn't really have functions like this one

apaszke commented on torch/csrc/jit/script/module.h:
```diff
@@ -105,124 +122,133 @@ struct NamedParameter {
+struct OrderedDict {
+  OrderedDict() {}
+  T& insert(const std::string& name,  T&& value, const char* what) {
+    if(index_.count(name) != 0) {
```
FWIW `OrderedDict`s support insertion (items stay in the same order, only value is updated). Don't know if this is a general constraint coming from Modules, or you didn't implement it for some other reason

apaszke commented on torch/csrc/jit/script/module.h:
```diff
@@ -105,124 +122,133 @@ struct NamedParameter {
+      graph = std::make_shared<Graph>();
+    std::unique_ptr<Method> method(new Method(name, optimize, std::move(graph), std::move(member_inputs)));
+
+    return *methods.insert(name, std::move(method), "method");
```
We don't need an *ordered* dict for methods right?

apaszke replied:
Also, it looks like the `what` arguments are always the same for any given dict. Can't we make them a value given to the constructor, so we can avoid passing them in every time?

apaszke commented on torch/csrc/jit/script/module.h:
```diff
@@ -105,124 +122,133 @@ struct NamedParameter {
-  std::unordered_map<std::string, NamedMember> members;
+  OrderedDict<NamedModule> modules;
+  OrderedDict<NamedParameter> parameters;
+  OrderedDict<std::unique_ptr<Method>> methods;
```
Can you remind me why do we keep `unique_ptr`s here? Is it because `methods` are quite heavyweight and we want to avoid the cost of moving?

zdevito replied:
Methods are owned by their  module, but we return them as references that are held by python (and tied to the lifetime of the Module). We need to ensure they do not get relocated (e.g. by resizing the vector).

apaszke commented on torch/jit/__init__.py:
```diff
@@ -463,149 +463,14 @@ def wrapper(func):
         for name in executor_options:
             executor_options[name] = kwargs.pop(name, executor_options[name])
```
Can you add a check that `kwargs` are empty after this loop? I just noticed we forgot about it, and will silently ignore anything we don't support (e.g. names with typos)

apaszke commented on torch/jit/__init__.py:
```diff
@@ -463,149 +463,14 @@ def wrapper(func):
-            return TracedModule(func, executor)
+            module = TracedModule(func, **executor_options)
+            module._create_method_from_trace('forward', func, args)
+            return module
```
Have you benchmarked this change?

zdevito replied:
No, it would be a good idea to do, but I want to unblock @jamesr66a first. I think conceptually the timing will not change just because we load parameters via pointers, but it is of course possible this adds some other overhead, so timing is a good idea.

apaszke commented on torch/jit/__init__.py:
```diff
@@ -688,6 +667,7 @@ def __init__(cls, name, bases, attrs):
             if isinstance(v, ScriptMethodStub):
                 delattr(cls, k)
                 methods.append(v)
+        # prevent messing with original script module
```
Can you elaborate?

zdevito replied:
stale comment, removed

apaszke commented on torch/jit/__init__.py:
```diff
@@ -702,32 +682,100 @@ def init_then_register(self, *args, **kwargs):
+class ScriptModule(with_metaclass(ScriptMeta, Module, torch._C.ScriptModule)):
+    def __init__(self, optimize=True):
+        Module.__init__(self)
+        torch._C.ScriptModule.__init__(self, optimize)
```
We should make sure that forgetting to call `super().__init__()` in a user-defined subclass can't cause a segfault. I've seen it happen multiple times with `pybind`, because it doesn't construct the C++ object before `__init__` (it should probably do that in `__new__`, but oh well)

zdevito replied:
I moved the call to ._C.ScriptModule into the init that is generated by the metaclass, so it gets called even if super is not called.

apaszke commented on torch/jit/__init__.py:
```diff
@@ -702,32 +682,100 @@ def init_then_register(self, *args, **kwargs):
-            raise AttributeError("'{}' object has no attribute '{}'".format(self.__class__.__name__, attr))
-        return r
+        if self._has_method(attr):
+            return self._get_method(attr)
```
Not very important: would be also nice to add methods to `__dir__`

apaszke commented on torch/jit/__init__.py:
```diff
@@ -669,6 +534,119 @@ def script_method(fn):
+        return [v for k, v in self.items()]
+
+    def __delitem__(self, k):
+        raise RuntimeError("cannot delete methods or parameters of a script module")
```
Can you mention buffers too? Otherwise someone might get this message when trying to remove a buffer

apaszke commented on torch/jit/__init__.py:
```diff
@@ -669,6 +534,119 @@ def script_method(fn):
+
+    def items(self):
+        r = self.module._get_modules()
+        return r
```
nit: Unnecessary assignment

apaszke commented on torch/jit/__init__.py:
```diff
@@ -669,6 +534,119 @@ def script_method(fn):
+        # to ensure that you always get the same python value here
+        # we store it as a native attribute _in addition to_
+        # registering it with the C++ script::Module
+        object.__setattr__(self.module, k, v)
```
It looks like this is unnecessary or broken. We never access attributes of this dict object like we did when this was in module

apaszke commented on torch/jit/__init__.py:
```diff
@@ -669,6 +534,119 @@ def script_method(fn):
+        object.__setattr__(self.module, k, v)
+
+    def __getitem__(self, k):
+        return self._get_module(k)
```
Forgotten `.module`?

apaszke replied:
Also missing a `KeyError`

apaszke commented on torch/jit/__init__.py:
```diff
@@ -702,32 +682,100 @@ def init_then_register(self, *args, **kwargs):
+    # Module already has this method defined, so we
+    # need to override it and send it through the ScriptModule lookup
+    def forward(self, *args, **kwargs):
+        return self.__getattr__('forward')(*args, **kwargs)
```
Using `__getattr__` is a bit conterintuitive. Can't you just `self._get_method('forward')`? That will be more explicit and faster

apaszke commented on torch/jit/__init__.py:
```diff
@@ -702,32 +682,100 @@ def init_then_register(self, *args, **kwargs):
+    if name.startswith('__'):
+        continue
+    if name not in ScriptModule.__dict__ and name not in _compiled_methods_whitelist:
+        setattr(ScriptModule, method.__name__, _make_fail(name))
```
Not sure if we want an equally strict whitelist in `ScriptModule`. It was written with `TracedModule` in mind, so it really includes only "`const` methods".

zdevito replied:
I did go through and  edit the whitelist to set it to the things I think would work given the current code in `nn.Module`, if there are specifics functions we can add that are not in this list, we should.

apaszke reviewed (approved):
LGTM (long review so remember to expand the comment list).

A more complete set of tests would probably need the same things that a list of supported features would have (would be good to write this up too). An incomplete list would include:
- having non-scripted methods
- having scripted methods in arbitrary order
- having non-parameter/module/method attributes
- allowed changes to attributes inside and outside of `__init__` including at least adding/removing/changing params/buffers/submodules/regular Python attrs
- calling submodules/(non-forward scripted/non-scripted methods)

zdevito committed:
PR comments

zdevito commented on test/test_jit.py:
```diff
@@ -1785,12 +1785,11 @@ def forward(self, x):
 
         # Submodules can't be called
-        with self.assertRaises(RuntimeError):
+        with self.assertRaises(AttributeError):
```
I changed this back to RuntimeError by giving a clearer error message.

zdevito commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -155,6 +153,40 @@ py::object unpackVariableTensorList(std::vector<at::Tensor> outputs) {
+  return graph;
+}
+
+static void gatherStateDict(std::vector<at::Tensor*> & values, const Module & m) {
```
It includes both buffers and parameters, since `parameters` in C++ includes both. I updated the name to make this clear. Deduplication is handled in python when constructing the trace modules in the first place.

apaszke replied:
Can we document the invariant that says that params are always deduplicated (including occurence in other modules) in C++?

zdevito committed:
PR comments

zdevito commented:
I addressed most of the comments.  I left a few things for future PR to test and/or implement:
* support non-script methods
* support non-script attributes
* change non-script modules after creation
* change inlining resolution order so that methods can be defined in arbitrary order.

apaszke commented:
Should be good to merge once the tests are green. What does "change non-script modules after creation" mean in this case?

apaszke commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -297,18 +270,25 @@ void initJitScriptBindings(PyObject* module) {
+        for(auto& m : self.get_methods()) {
+          r.push_back(m->name());
+        }
+        return r;
```
nit: `fmap` would simplify this

apaszke commented on torch/jit/__init__.py:
```diff
@@ -462,6 +462,9 @@ def wrapper(func):
         for name in executor_options:
             executor_options[name] = kwargs.pop(name, executor_options[name])
+        if len(kwargs) != 0:
+            raise TypeError("got unexpected keyword arguments: {}".format(",".join(kwargs.keys())))
```
nit: it's better to use `', '.join(...)` (the space after the comma improves readability)

apaszke commented on torch/jit/__init__.py:
```diff
@@ -772,6 +782,11 @@ def check_unique(param):
+    def forward(self, *args, **kwargs):
+        if self._has_method('forward'):
+            return self._get_method('forward')(*args, **kwargs)
+        raise RuntimeError('Trace submodules cannot be called.')
```
Would be faster to use `_get_method` in `try ... except` and raise a readable error if this returns a KeyError or sth like this

apaszke reviewed (commented):
Some minor suggestions. Would be good to improve at least `ScriptModule.forward` (other are completely minor)

zdevito committed:
PR Comments

zdevito committed:
build fix

zdevito committed:
build fixes

zdevito committed:
remove has_method check from TracedModule

zdevito commented:
@onnxbot retest this please
